cot/src/email.rs

```
@@ -0,0 +1,590 @@
+    /// creating the TLS parameters, or connecting to the SMTP server.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the transport is not properly initialized.
+    fn open(&mut self) -> Result<()> {
+        // Test if self.transport is None or if the connection is not working
+        if self.transport.is_some() && self.transport.as_ref().unwrap().test_connection().is_ok() {
+            return Ok(());
+        }
+        if self.config.host.is_empty() {
+            return Err(EmailError::ConfigurationError(
+                "SMTP host is required".to_string(),
+            ));
+        } else if self.config.port == 0 {
+            return Err(EmailError::ConfigurationError(
+                "SMTP port is required".to_string(),
+            ));
+        }
+        let _socket_addr = format!("{}:{}", self.config.host, self.config.port)
+            .to_socket_addrs()
+            .map_err(|e| EmailError::ConnectionError(e.to_string()))?
+            .next()
+            .ok_or_else(|| {
+                EmailError::ConnectionError("Could not resolve SMTP host".to_string())
+            })?;
+
+        let mut transport_builder = SmtpTransport::builder_dangerous(&self.config.host)
+            .port(self.config.port)
+            .timeout(Some(self.config.timeout));
+
+        // Add authentication if credentials provided
+        if let (Some(username), Some(password)) = (&self.config.username, &self.config.password) {
+            let credentials = Credentials::new(username.clone(), password.clone());
+            transport_builder = transport_builder.credentials(credentials);
+        }
+
+        // Connect to the SMTP server
+        let transport = transport_builder.build();
+        if transport.test_connection().is_err() {
+            return Err(EmailError::ConnectionError(
+                "Failed to connect to SMTP server".to_string(),
+            ));
+        }
```

Please bubble up the error from `test_connection()`, eg.
```rust
let error_message = format!("Failed to connect to SMTP server: {e}";
return Err(EmailError::ConnectionError(error_message.to_string()));
```